optuna/integration/lightgbm_tuner/sklearn.py

```diff
@@ -1,58 +1,1294 @@
+        self._boosters = env.model.boosters
+
+
+class _Objective(object):
+    def __init__(
+        self,
+        params: Dict[str, Any],
+        dataset: lgb.Dataset,
+        eval_name: str,
+        is_higher_better: bool,
+        n_samples: int,
+        callbacks: Optional[List[Callable]] = None,
+        categorical_feature: Union[List[int], List[str], str] = "auto",
+        cv: Optional[CVType] = None,
+        early_stopping_rounds: Optional[int] = None,
+        enable_pruning: bool = False,
+        feature_name: Union[List[str], str] = "auto",
+        feval: Optional[Callable] = None,
+        fobj: Optional[Callable] = None,
+        init_model: Optional[Union[lgb.Booster, lgb.LGBMModel, str]] = None,
+        n_estimators: int = 100,
+        param_distributions: Optional[Dict[str, distributions.BaseDistribution]] = None,
+    ) -> None:
+        self.callbacks = callbacks
+        self.categorical_feature = categorical_feature
+        self.cv = cv
+        self.dataset = dataset
+        self.early_stopping_rounds = early_stopping_rounds
+        self.enable_pruning = enable_pruning
+        self.eval_name = eval_name
+        self.feature_name = feature_name
+        self.feval = feval
+        self.fobj = fobj
+        self.init_model = init_model
+        self.is_higher_better = is_higher_better
+        self.n_estimators = n_estimators
+        self.n_samples = n_samples
+        self.params = params
+        self.param_distributions = param_distributions
+
+    def __call__(self, trial: trial_module.Trial) -> float:
+        params = self._get_params(trial)  # type: Dict[str, Any]
+        dataset = copy.copy(self.dataset)
+        callbacks = self._get_callbacks(trial)  # type: List[Callable]
+        eval_hist = lgb.cv(
+            params,
+            dataset,
+            callbacks=callbacks,
+            categorical_feature=self.categorical_feature,
+            early_stopping_rounds=self.early_stopping_rounds,
+            feature_name=self.feature_name,
+            feval=self.feval,
+            fobj=self.fobj,
+            folds=self.cv,
+            init_model=self.init_model,
+            num_boost_round=self.n_estimators,
+        )  # Dict[str, List[float]]
+        best_iteration = callbacks[0]._best_iteration  # type: ignore
+
+        trial.set_user_attr("best_iteration", best_iteration)
+
+        value = eval_hist["{}-mean".format(self.eval_name)][-1]  # type: float
+        is_best_trial = True  # type: bool
+
+        try:
+            is_best_trial = (value < trial.study.best_value) ^ self.is_higher_better
+        except ValueError:
+            pass
+
+        if is_best_trial:
+            boosters = callbacks[0]._boosters  # type: ignore
+            representations = []  # type: List[str]
+
+            for b in boosters:
+                b.free_dataset()
+                representations.append(b.model_to_string())
+
+            trial.study.set_user_attr("representations", representations)
```

Comment:
This implementation stores boosting models in optuna's storage.
SQLite can store up to 1 billion bytes in a row of string fields by default[1].
In RDB (MySQL), on the other hand, the maximum row field size of `TrialUserAttributeModel` is 2048 bytes[2].
I think this is too small to save the boosting model. What do you think about this?
Is it possible to do parallel execution in a practical case?

----
[1] https://www.sqlite.org/limits.html
[2] https://github.com/optuna/optuna/blob/master/optuna/storages/rdb/models.py#L276-L282

Reply:
Thank you for pointing this out. I did not consider at all.

In the example below, boosters consume more than 10 MB of memory.
I can't think of a solution other than to save boosters to files instead.

```python
def total_size(o, handlers={}, verbose=False):
    """ Returns the approximate memory footprint an object and all of its contents.

    Automatically finds the contents of the following builtin containers and
    their subclasses:  tuple, list, deque, dict, set and frozenset.
    To search other containers, add handlers to iterate over their contents:

        handlers = {SomeContainerClass: iter,
                    OtherContainerClass: OtherContainerClass.get_elements}

    """
    # See https://code.activestate.com/recipes/577504/
    ...


import optuna.integration.lightgbm as lgb
from sklearn.datasets import load_digits

X, y = load_digits(return_X_y=True)

clf = lgb.LGBMClassifier(n_estimators=100, random_state=0)
clf.fit(X, y, early_stopping_rounds=None)
total_size(clf.study_.user_attrs["representations"])  # 10743344 bytes

clf = lgb.LGBMClassifier(n_estimators=1000, random_state=0)
clf.fit(X, y, early_stopping_rounds=None)
total_size(clf.study_.user_attrs["representations"])  # 19605949 bytes
```

Reply:
@Y-oHr-N, thank you for continuing to look at this PR!
I tested the almost same code with using RDB storage (MySQL). The error message is here: https://gist.github.com/smly/ec26cff6dec17568d97006a573dda5f8
As for supporting parallel execution, I think the discussion on #1076 will help resolve the issue. It is ongoing PR, but IMO (1) and (3) are preferable on practical usage. Your comments or insights for this problem are welcome!

Reply:
I fixed in Y-oHr-N/OptGBM#74 to save the booster as a file. If you like it, I'll fix it in a similar way.

Reply:
Thank you for the updates, I will check the changes.

Reply:
In parallel execution, different processes may write to the same file name at the same time in `dump()` method. This will corrupt the output file.

For example, if another process updates the best score while a process is outputting a file, multiple processes will output the same file. Is my understanding correct?

Reply:
You are right. Here are my solutions to address this issue:

(1) Lock a file while writting. (Exclusive control can be achieved using [fcntl](https://docs.python.org/3/library/fcntl.html), but it is not available on Windows)
(2) Separate directories by the trial number. (In this case, boosters are not overwitten)

There are also two patterns for saving boosters.

(a) Save boosters at the end of each trial
(b) Save boosters only if the best score is updated

#1076 employs (2) (trial number in the file name) and (a). These result in increased writing time and file size.

Reply:
It is quite reasonable not to save all the boosters.
IMO, `fcntl` is a good way to implement exclusive write even though Windows is not supported. For supporting Windows, naive file locking is also an option to implement it.

Let me have some time to think about this. I would like to hear ideas/opinions from other committers.

Reply:
Sorry for my late review. I will resume the review process from today. I also asked one of the other committers to review. Especially about parallelization. Thanks.

Reply:
Using `fcntl` is not a good option, especially when using this Tuner with NFS, as NFS is notorious for broken lock implementations. If storage consumption is a serious issue, you can add a storage compression mechanism via study callbacks, but I think it should be addressed in another PR.

Reply:
Thanks, @ytsmiling. Based on your review, I think it would be preferable to implement it in the way of (2)(a) and fix it in another PR.

@smily, do you have any suggestions?

Reply:
@ytsmiling, thanks for your review! Totally agree with your suggestion. I think your idea of using callback for storage compression is great.

@Y-oHr-N, it looks good to me. Thank you for continuing to look at this PR.